File: backend/server/models/transaction.go
package models

type Transaction struct {
	// The ID of the transaction.
	//
	// Will be ignored when writing.
	ID                 int64   `json:"id"`
	Date               string  `json:"date" binding:"required,iso8601"`
	Account            string  `json:"account" binding:"required,max=20"`
	VerificationNumber string  `json:"verificationNumber" binding:"max=20"`
	Text               string  `json:"text" binding:"required,max=45"`
	Description        string  `json:"description" binding:"max=255"`
	Amount             float64 `json:"amount" binding:"required"`
	CategoryId         *int64  `json:"categoryId"`
}

type TransactionAddBody struct {
	Transactions []Transaction `json:"transactions" binding:"required"`
}

type TransactionGetListBody struct {
	TotalCount   int64         `json:"totalCount"`
	Transactions []Transaction `json:"transactions"`
}

type ResultByMonth struct {
	Year     int64   `json:"year"`
	Month    int64   `json:"month"`
	Income   float64 `json:"income"`
	Expenses float64 `json:"expenses"`
	Internal bool    `json:"internal"`
}

File: expensetrack/lib/models/transaction.dart
import 'package:expensetrack/app/json_helpers.dart';
import 'package:freezed_annotation/freezed_annotation.dart';

part 'transaction.freezed.dart';
part 'transaction.g.dart';

@freezed
class Transaction with _$Transaction {
  factory Transaction({
    /// The ID of the category.
    ///
    /// Will be ignored when writing.
    int? id,

    /// The date that the transaction was made.
    ///
    /// Must be formatted according to the ISO 8601 standard.
    @JsonKey(fromJson: JsonHelpers.dateFromJson, toJson: JsonHelpers.dateToJson)
    required DateTime date,

    /// The account number that the transaction belongs to.
    ///
    /// Max 20 characters.
    required String account,
    
    /// The verification number of the transaction.
    ///
    /// Max 20 characters.
    String? verificationNumber,

    /// The Text from the bank for the transaction.
    /// 
    /// Max 45 characters.
    required String text,

    /// A description of the transaction.
    /// 
    /// Max 255 characters.
    String? description,

    required num amount,
    
    /// The id of the category that the transaction belongs to.
    int? categoryId,
  }) = _Transaction;

  factory Transaction.fromJson(Map<String, dynamic> json) => _$TransactionFromJson(json);
}

@freezed
class GetTransactionsBody with _$GetTransactionsBody {
  factory GetTransactionsBody({
    required int totalCount,
    required List<Transaction> transactions,
  }) = _GetTransactionsBody;

factory GetTransactionsBody.fromJson(Map<String, dynamic> json) => 
_$GetTransactionsBodyFromJson(json);
}
